// @appium/css-locator-to-native/build/lib/internal/normalize-from-parsed.js
import { UnsupportedSelectorError } from '../public/errors.js';
import { isEmpty } from './utils.js';
const SUPPORTED_COMBINATORS = new Set([' ', '>', undefined]);
/**
 * Normalize an internally parsed CSS selector into the public {@link ParsedSelector} IR.
 *
 * @param parsed - Parser output in the internal selector shape
 * @param schema - Caller-defined attribute vocabulary and coercion rules
 * @returns Normalized selector intermediate representation (first rule only)
 * @throws {UnsupportedSelectorError} If the parsed selector uses unsupported features or attributes
 */
export function normalizeFromParsed(parsed, schema) {
    if (isEmpty(parsed.rules)) {
        throw new UnsupportedSelectorError('No rules could be parsed out of the current selector');
    }
    const aliasMap = buildAliasMap(schema);
    return {
        rule: normalizeRule(parsed.rules[0], schema, aliasMap),
    };
}
function buildAliasMap(schema) {
    const aliases = new Map();
    for (const [canonicalName, definition] of Object.entries(schema.attributes)) {
        aliases.set(canonicalName.toLowerCase(), canonicalName);
        for (const alias of definition.aliases ?? []) {
            aliases.set(alias.toLowerCase(), canonicalName);
        }
    }
    return aliases;
}
function resolveEntityName(name, schema, aliasMap) {
    const canonical = aliasMap.get(name.toLowerCase());
    if (!canonical) {
        throw new UnsupportedSelectorError(`'${name}' is not a valid attribute. Supported attributes are: '${Object.keys(schema.attributes).join(', ')}'`);
    }
    return canonical;
}
function coerceBoolean(rawValue, entityName, booleanFormat) {
    const format = booleanFormat ?? 'literal';
    if (format === 'literal') {
        return rawValue;
    }
    const value = (rawValue ?? '').toLowerCase() || 'true';
    switch (format) {
        case 'zero-one':
            switch (value) {
                case '0':
                case 'false':
                    return '0';
                case '1':
                case 'true':
                    return '1';
                default:
                    throw new UnsupportedSelectorError(`'${entityName}' must be true/1 or false/0 or empty. Found '${rawValue ?? ''}'`);
            }
        case 'true-false':
            switch (value) {
                case 'true':
                    return 'true';
                case 'false':
                    return 'false';
                default:
                    throw new UnsupportedSelectorError(`'${entityName}' must be true, false or empty. Found '${rawValue ?? ''}'`);
            }
        default:
            throw new UnsupportedSelectorError(`Unsupported boolean format '${booleanFormat}'.`);
    }
}
function assertStringValue(value, entityName) {
    if (!isEmpty(value) && typeof value !== 'string') {
        throw new UnsupportedSelectorError(`'${entityName}=${value}' is an invalid attribute. Only string and empty attribute types are supported.`);
    }
}
function normalizeEntity(name, value, operator, schema, aliasMap) {
    const canonicalName = resolveEntityName(name, schema, aliasMap);
    const definition = schema.attributes[canonicalName];
    if (!definition) {
        throw new UnsupportedSelectorError(`'${name}' is not a valid attribute.`);
    }
    assertStringValue(value, canonicalName);
    switch (definition.type) {
        case 'boolean': {
            const coerced = coerceBoolean(value, canonicalName, schema.booleanFormat);
            const attr = {
                name: canonicalName.toLowerCase(),
                implicit: value === undefined,
            };
            if (coerced !== undefined) {
                attr.value = coerced;
            }
            return attr;
        }
        case 'numeric': {
            const attr = {
                name: canonicalName.toLowerCase(),
                value,
            };
            if (operator) {
                attr.operator = operator;
            }
            return attr;
        }
        case 'string':
            if (operator && !['=', '*=', '^=', '$=', '~='].includes(operator)) {
                throw new UnsupportedSelectorError(`Unsupported CSS attribute operator '${operator}'.`);
            }
            return {
                name: canonicalName.toLowerCase(),
                operator,
                value: value ?? '',
            };
        default:
            throw new UnsupportedSelectorError(`Unsupported attribute type for '${canonicalName}'.`);
    }
}
function normalizeAttribute(attr, schema, aliasMap) {
    return normalizeEntity(attr.name, attr.value, attr.operator, schema, aliasMap);
}
function normalizePseudo(pseudo, schema, aliasMap) {
    return normalizeEntity(pseudo.name, pseudo.value, undefined, schema, aliasMap);
}
function mapCombinator(combinator) {
    if (!combinator || combinator === ' ') {
        return 'descendant';
    }
    if (combinator === '>') {
        return 'child';
    }
    throw new UnsupportedSelectorError(`'${combinator}' is not a supported combinator. Only child combinator (>) and descendant combinator are supported.`);
}
function normalizeRule(rule, schema, aliasMap, isNested = false) {
    if (rule.combinator && !SUPPORTED_COMBINATORS.has(rule.combinator)) {
        throw new UnsupportedSelectorError(`'${rule.combinator}' is not a supported combinator. Only child combinator (>) and descendant combinator are supported.`);
    }
    const parsedRule = {
        classes: [...rule.classNames],
        attributes: rule.attributes.map((attr) => normalizeAttribute(attr, schema, aliasMap)),
        pseudos: rule.pseudoClasses.map((pseudo) => normalizePseudo(pseudo, schema, aliasMap)),
    };
    if (rule.tag !== undefined) {
        parsedRule.tag = rule.tag;
    }
    if (rule.ids.length) {
        parsedRule.id = rule.ids[0];
    }
    if (isNested || rule.combinator) {
        parsedRule.combinator = mapCombinator(rule.combinator);
    }
    if (rule.nestedRule) {
        parsedRule.nested = normalizeRule(rule.nestedRule, schema, aliasMap, true);
    }
    return parsedRule;
}
