// appium-uiautomator2-driver/build/lib/css/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UI_AUTOMATOR_STRATEGY = void 0;
exports.cssToNativeLocator = cssToNativeLocator;
const driver_1 = require("appium/driver");
const utils_1 = require("../utils");
const constants_1 = require("./constants");
const schema_1 = require("./schema");
const ui_automator_emitter_1 = require("./ui-automator-emitter");
var constants_2 = require("./constants");
Object.defineProperty(exports, "UI_AUTOMATOR_STRATEGY", { enumerable: true, get: function () { return constants_2.UI_AUTOMATOR_STRATEGY; } });
const emitters = {
    [constants_1.UI_AUTOMATOR_EMITTER_KEY]: new ui_automator_emitter_1.UiAutomatorEmitter(constants_1.UI_AUTOMATOR_STRATEGY),
};
const getTransformCss = (0, utils_1.memoize)(async function loadTransformCss() {
    const mod = await import('@appium/css-locator-to-native');
    return mod.createCssTransformer({
        schema: schema_1.ATTRIBUTE_SCHEMA,
        emitters,
        resolveStrategy() {
            return constants_1.UI_AUTOMATOR_EMITTER_KEY;
        },
    });
});
/**
 * Converts a CSS selector string into a native locator for the resolved strategy.
 *
 * @param css - CSS selector to transform
 * @param appPackage - Optional application package used to qualify resource ids
 * @returns Native locator strategy name and selector string
 */
async function cssToNativeLocator(css, appPackage) {
    try {
        const transformCss = await getTransformCss();
        return transformCss(css, { appPackage });
    }
    catch (err) {
        throw mapCssError(err, css);
    }
}
function mapCssError(err, css) {
    if (isPackageError(err, 'InvalidSelectorError')) {
        return new driver_1.errors.InvalidSelectorError(`Invalid CSS selector '${css}'`, err);
    }
    if (isPackageError(err, 'UnsupportedSelectorError')) {
        return new driver_1.errors.InvalidSelectorError(`Unsupported CSS selector '${css}'`, err);
    }
    if (err instanceof Error) {
        return err;
    }
    return new Error(String(err));
}
function isPackageError(err, name) {
    return err instanceof Error && err.name === name;
}
