// appium-uiautomator2-driver/build/lib/css/ui-automator-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UiAutomatorEmitter = void 0;
const utils_1 = require("../utils");
const BOOLEAN_ATTRS = new Set([
    'checkable',
    'checked',
    'clickable',
    'enabled',
    'focusable',
    'focused',
    'long-clickable',
    'scrollable',
    'selected',
]);
const NUMERIC_ATTRS = new Set(['index', 'instance']);
const STRING_ATTRS = new Set(['description', 'resource-id', 'text', 'class-name', 'package-name']);
const ID_LOCATOR_PATTERN = /^[a-zA-Z_][a-zA-Z0-9._]*:id\/[\S]+$/;
/** Converts parsed CSS selectors into UiAutomator selector strings. */
class UiAutomatorEmitter {
    strategy;
    constructor(strategy) {
        this.strategy = strategy;
    }
    emit(parsed, ctx) {
        return this.emitRule(parsed.rule, ctx);
    }
    emitRule(rule, ctx) {
        const parts = ['new UiSelector()'];
        const tagName = rule.tag;
        const classNames = [...rule.classes];
        if (tagName && tagName !== '*') {
            if (classNames.length) {
                parts.push(`.className("${[tagName, ...classNames].join('.')}")`);
            }
            else {
                parts.push(`.classNameMatches("${tagName}")`);
            }
        }
        else if (classNames.length) {
            parts.push(`.classNameMatches("${classNames.join('\\.')}")`);
        }
        if (rule.id) {
            parts.push(`.resourceId("${this.formatIdLocator(rule.id, ctx?.appPackage)}")`);
        }
        for (const attr of rule.attributes) {
            parts.push(this.formatEntity(attr, ctx));
        }
        for (const pseudo of rule.pseudos) {
            const formatted = this.formatEntity(pseudo, ctx);
            if (formatted) {
                parts.push(formatted);
            }
        }
        if (rule.nested) {
            parts.push(`.childSelector(${this.emitRule(rule.nested, ctx)})`);
        }
        return parts.join('');
    }
    formatEntity(attr, ctx) {
        if (BOOLEAN_ATTRS.has(attr.name)) {
            return this.formatBoolean(attr);
        }
        if (NUMERIC_ATTRS.has(attr.name)) {
            return `.${attr.name}(${attr.value})`;
        }
        if (STRING_ATTRS.has(attr.name)) {
            return this.formatStringAttr(attr, ctx);
        }
        return '';
    }
    formatBoolean(attr) {
        const value = attr.value ?? 'true';
        return `.${toSnakeCase(attr.name)}(${value})`;
    }
    formatStringAttr(attr, ctx) {
        const methodName = toSnakeCase(attr.name);
        let value = attr.value ?? '';
        if (attr.name === 'resource-id') {
            value = this.formatIdLocator(value, ctx?.appPackage);
        }
        if (value === '') {
            return `.${methodName}Matches("")`;
        }
        switch (attr.operator) {
            case '=':
                return `.${methodName}("${value}")`;
            case '*=':
                if (['description', 'text'].includes(attr.name)) {
                    return `.${methodName}Contains("${value}")`;
                }
                return `.${methodName}Matches("${(0, utils_1.escapeRegExp)(value)}")`;
            case '^=':
                if (['description', 'text'].includes(attr.name)) {
                    return `.${methodName}StartsWith("${value}")`;
                }
                return `.${methodName}Matches("^${(0, utils_1.escapeRegExp)(value)}")`;
            case '$=':
                return `.${methodName}Matches("${(0, utils_1.escapeRegExp)(value)}$")`;
            case '~=':
                return `.${methodName}Matches("${getWordMatcherRegex(value)}")`;
            default:
                return `.${methodName}("${value}")`;
        }
    }
    formatIdLocator(locator, appPackage) {
        return ID_LOCATOR_PATTERN.test(locator) ? locator : `${appPackage || 'android'}:id/${locator}`;
    }
}
exports.UiAutomatorEmitter = UiAutomatorEmitter;
function toSnakeCase(str) {
    const tokens = str
        .split('-')
        .map((token) => token.charAt(0).toUpperCase() + token.slice(1).toLowerCase());
    const out = tokens.join('');
    return out.charAt(0).toLowerCase() + out.slice(1);
}
function getWordMatcherRegex(word) {
    return `\\b(\\w*${(0, utils_1.escapeRegExp)(word)}\\w*)\\b`;
}
