// @appium/base-driver/build/lib/basedriver/ipc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppiumIpc = exports.IpcSubscription = exports.EVT_UNSUBSCRIBED = exports.EVT_MESSAGE = void 0;
const logger_1 = require("./logger");
const node_events_1 = __importDefault(require("node:events"));
const asyncbox_1 = require("asyncbox");
const support_1 = require("@appium/support");
const DEF_MAX_OBJ_SIZE_BYTES = 1024 * 1024; // 1mb seems like plenty for any plugin to pass a message
const DEF_MAX_TOPICS = 1000;
exports.EVT_MESSAGE = 'message';
exports.EVT_UNSUBSCRIBED = 'unsubscribed';
const ASYNC_ITERATOR_STOP = Symbol('asyncIteratorStop');
class IpcSubscription extends node_events_1.default {
    subscriber;
    topic;
    ipc;
    constructor(subscriber, topic, ipc) {
        super();
        this.subscriber = subscriber;
        this.topic = topic;
        this.ipc = ipc;
    }
    get isActive() {
        return this.ipc.subscriptionExists(this.topic, this.subscriber);
    }
    getMessage() {
        if (!this.isActive) {
            throw new Error('Cannot get message from subscription after unsubscribing');
        }
        return this.ipc.getMessage(this.topic);
    }
    async publish(data) {
        if (!this.isActive) {
            throw new Error('Cannot publish data to topic from subscription after unsubscribing');
        }
        return await this.ipc.publish(this.topic, this.subscriber, data);
    }
    unsubscribe() {
        if (!this.isActive) {
            return false;
        }
        const unsubscribeRes = this.ipc.unsubscribe(this.topic, this.subscriber);
        this.emit('unsubscribed');
        this.removeAllListeners(exports.EVT_MESSAGE);
        return unsubscribeRes;
    }
    async *[Symbol.asyncIterator]() {
        // yield any messages that are emitted, but keep an eye out for an unsubscribed that happens
        // while a caller is waiting on the loop, because we want to exit the loop in case of
        // unsubscription, even if we were already waiting on the next message.
        while (this.isActive) {
            const val = await new Promise((resolve) => {
                this.once(exports.EVT_MESSAGE, (message) => {
                    this.removeAllListeners(exports.EVT_UNSUBSCRIBED);
                    resolve(message);
                });
                this.once(exports.EVT_UNSUBSCRIBED, () => {
                    // EVT_MESSAGE listeners are already removed in unsubscribe()
                    resolve(ASYNC_ITERATOR_STOP);
                });
            });
            if (val === ASYNC_ITERATOR_STOP) {
                break;
            }
            yield val;
        }
    }
}
exports.IpcSubscription = IpcSubscription;
class AppiumIpc {
    messageByTopic = {};
    subs = {};
    topics = new Set();
    maxObjSize;
    maxTopics;
    log;
    constructor(opts = {}) {
        this.maxObjSize = opts.maxObjSize ?? DEF_MAX_OBJ_SIZE_BYTES;
        this.maxTopics = opts.maxTopics ?? DEF_MAX_TOPICS;
        this.log = opts.log ?? logger_1.log;
        this.log.debug(`Initialized new IPC object with max object size of ${this.maxObjSize} bytes ` +
            `and max topics of ${this.maxTopics}`);
    }
    subscribe(topic, subscriber) {
        this.log.info(`Subscribing ${subscriber} to topic '${topic}'`);
        if (this.subscriptionExists(topic, subscriber)) {
            throw new Error(`Subscription already exists for topic "${topic}" and subscriber "${subscriber}"`);
        }
        this.ensureTopic(topic);
        this.subs[topic] ??= [];
        const sub = new IpcSubscription(subscriber, topic, this);
        this.subs[topic].push(sub);
        return sub;
    }
    unsubscribe(topic, subscriber) {
        this.log.info(`Unsubscribing ${subscriber} from topic '${topic}'`);
        if (this.subscriptionExists(topic, subscriber)) {
            this.subs[topic] = this.subs[topic].filter((sub) => sub.subscriber !== subscriber);
            return true;
        }
        return false;
    }
    async publish(topic, publisher, data) {
        this.log.debug(`${publisher} is publishing a message to topic ${topic}`);
        this.ensureTopic(topic);
        const messageSize = support_1.node.getObjectSize(data);
        if (messageSize > this.maxObjSize) {
            throw new Error(`Error when ${publisher} is publishing to topic '${topic}': ` +
                `Message with size ${messageSize} bytes is bigger than max size of ${this.maxObjSize} bytes`);
        }
        let clonedData;
        try {
            clonedData = structuredClone(data);
        }
        catch (e) {
            throw new Error(`Could not clone data for IPC publish from ${publisher} on topic ${topic}`, { cause: e });
        }
        const message = { publisher, data: clonedData, topic, timestampMs: Date.now() };
        this.messageByTopic[topic] = message;
        const subs = this.subs[topic] ?
            this.subs[topic].filter((sub) => sub.subscriber !== publisher) :
            [];
        for (const sub of subs) {
            sub.emit(exports.EVT_MESSAGE, structuredClone(message));
        }
        // we don't want to return from publish until the async iterators on subscriptions have had
        // a chance to observe the emitted value, otherwise some might get lost
        await (0, asyncbox_1.sleep)(0);
    }
    getMessage(topic) {
        if (!this.messageByTopic[topic]) {
            return;
        }
        return structuredClone(this.messageByTopic[topic]);
    }
    subscriptionExists(topic, subscriber) {
        return !!this.subs[topic]?.some((sub) => sub.subscriber === subscriber);
    }
    ensureTopic(topic) {
        if (this.topics.has(topic)) {
            return;
        }
        if (this.topics.size >= this.maxTopics) {
            throw new Error(`Cannot create new IPC topic '${topic}': ` +
                `maximum of ${this.maxTopics} topics per session reached. ` +
                `Adjust with the --max-ipc-topics server arg.`);
        }
        this.topics.add(topic);
    }
}
exports.AppiumIpc = AppiumIpc;
