// @appium/docutils/build/lib/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.argify = exports.isStringArray = void 0;
exports.relative = relative;
exports.stopwatch = stopwatch;
exports.kebabCase = kebabCase;
exports.mergeDefaultsDeep = mergeDefaultsDeep;
exports.spawnBackgroundProcess = spawnBackgroundProcess;
exports.execWithErrorHandling = execWithErrorHandling;
const node_child_process_1 = require("node:child_process");
const node_path_1 = __importDefault(require("node:path"));
const teen_process_1 = require("teen_process");
const support_1 = require("@appium/support");
function relative(from, to) {
    if (to === undefined) {
        return (nextTo) => `.${node_path_1.default.sep}${node_path_1.default.relative(from, nextTo)}`;
    }
    return `.${node_path_1.default.sep}${node_path_1.default.relative(from, to)}`;
}
/**
 * A stopwatch-like thing
 *
 * Used for displaying elapsed time in milliseconds
 * @param id - Unique identifier
 * @returns Function that returns the elapsed time in milliseconds
 */
function stopwatch(id) {
    const start = Date.now();
    stopwatch.cache.set(id, start);
    return () => {
        const result = Date.now() - (stopwatch.cache.get(id) ?? 0);
        stopwatch.cache.delete(id);
        return result;
    };
}
stopwatch.cache = new Map();
/**
 * Type guard to narrow an array to a string array
 * @param value any value
 * @returns `true` if the array is `string[]`
 */
const isStringArray = (value) => Array.isArray(value) && value.every((item) => typeof item === 'string');
exports.isStringArray = isStringArray;
/**
 * Converts an object of string values to an array of arguments for CLI
 *
 * Supports `boolean` and `number` values as well.  `boolean`s are assumed to be flags which default
 * to `false`, so they will only be added to the array if the value is `true`.
 */
const argify = (obj) => {
    const args = [];
    for (const [key, value] of Object.entries(obj)) {
        if (value === true) {
            args.push(`--${key}`);
        }
        else if (value === false || value === undefined) {
            continue;
        }
        else {
            args.push(`--${key}`, String(value));
        }
    }
    return args;
};
exports.argify = argify;
/**
 * Converts a string to kebab-case.
 * @param value Input string
 * @returns Kebab-cased string
 */
function kebabCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
        .replace(/[_\s]+/g, '-')
        .toLowerCase();
}
/**
 * Performs a deep "defaults" merge into a clone of `target`.
 * Only undefined properties in `target` are filled from `defaults`.
 * @param target Destination object
 * @param defaults Default values object
 * @returns Merged clone
 */
function mergeDefaultsDeep(target, defaults) {
    const out = structuredClone(target);
    const stack = [
        { dest: out, src: defaults },
    ];
    while (stack.length) {
        const next = stack.pop();
        if (!next) {
            continue;
        }
        const { dest, src } = next;
        for (const [key, srcVal] of Object.entries(src)) {
            const destVal = dest[key];
            if (destVal === undefined) {
                dest[key] = structuredClone(srcVal);
                continue;
            }
            if (support_1.util.isPlainObject(destVal) && support_1.util.isPlainObject(srcVal)) {
                stack.push({ dest: destVal, src: srcVal });
            }
        }
    }
    return out;
}
/**
 * Spawns a long-running "background" child process.  This is expected to only return control to the
 * parent process in the case of a nonzero exit code from the child process.
 * @param command Command to run
 * @param args Args to pass to command
 * @param opts Spawn options. `{stdio: 'inherit'}` will always be true
 * @privateRemarks `teen_process` is good for running a one-shot command, but not so great for
 * background tasks; we use node's `child_process` directly here to pass `stdio` through, since
 * `teen_process` basically does not respect `{stdio: 'inherit'}`.
 */
async function spawnBackgroundProcess(command, args, opts) {
    return new Promise((resolve, reject) => {
        (0, node_child_process_1.spawn)(command, args, { ...opts, stdio: 'inherit' })
            .on('error', reject)
            .on('close', (code) => {
            // can be null or number
            if (code) {
                return reject(new Error(`${command} exited with code ${code}`));
            }
            resolve();
        });
    });
}
/**
 * Wraps {@linkcode exec} with error handling that appends stderr to the thrown error message.
 */
async function execWithErrorHandling(cmd, args, opts) {
    try {
        return await (0, teen_process_1.exec)(cmd, args, opts);
    }
    catch (err) {
        const execErr = err;
        execErr.message = execErr.stderr ? `${execErr.message}\nCommand error:\n${execErr.stderr}` : execErr.message;
        throw execErr;
    }
}
