// @appium/support/build/lib/plist.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePlistFile = parsePlistFile;
exports.updatePlistFile = updatePlistFile;
exports.createBinaryPlist = createBinaryPlist;
exports.parseBinaryPlist = parseBinaryPlist;
exports.createPlist = createPlist;
exports.parsePlist = parsePlist;
const plist_1 = require("plist");
const bplist_creator_1 = __importDefault(require("bplist-creator"));
const bplist_parser_1 = require("bplist-parser");
const fs_1 = require("./fs");
const logger_1 = __importDefault(require("./logger"));
const util_1 = require("./util");
const BPLIST_IDENTIFIER = {
    BUFFER: Buffer.from('bplist00'),
    TEXT: 'bplist00',
};
/**
 * Parses a file in xml or binary format of plist
 *
 * @param plist - The plist file path
 * @param mustExist - If set to false, this method will return an empty object when the file doesn't exist
 * @param quiet - If set to false, the plist path will be logged in debug level
 * @returns Parsed plist as a JS object
 */
async function parsePlistFile(plist, mustExist = true, quiet = true) {
    if (!(await fs_1.fs.exists(plist))) {
        if (mustExist) {
            throw new Error(`Plist file doesn't exist: '${plist}'`);
        }
        if (!quiet) {
            logger_1.default.debug(`Plist file '${plist}' does not exist. Returning an empty plist.`);
        }
        return {};
    }
    let obj;
    let type = 'xml';
    try {
        const plistContent = await fs_1.fs.readFile(plist);
        obj = parsePlist(plistContent);
        if (BPLIST_IDENTIFIER.BUFFER.compare(plistContent, 0, BPLIST_IDENTIFIER.BUFFER.length) === 0) {
            type = 'binary';
        }
    }
    catch (err) {
        throw new Error(`Could not parse plist file '${plist}': ${err.message}`, { cause: err });
    }
    if (!quiet) {
        logger_1.default.debug(`Parsed plist file '${plist}' as ${type}`);
    }
    return obj;
}
/**
 * Updates a plist file with the given fields
 *
 * @param plist - The plist file path
 * @param updatedFields - The updated fields-value pairs
 * @param binary - If set to false, the file will be created as a xml plist
 * @param mustExist - If set to false, this method will update an empty plist
 * @param quiet - If set to false, the plist path will be logged in debug level
 */
async function updatePlistFile(plist, updatedFields, binary = true, mustExist = true, quiet = true) {
    let obj;
    try {
        obj = await parsePlistFile(plist, mustExist);
    }
    catch (err) {
        throw new Error(`Could not update plist: ${err.message}`, { cause: err });
    }
    Object.assign(obj, updatedFields);
    const newPlist = binary ? (0, bplist_creator_1.default)(obj) : (0, plist_1.build)(obj);
    try {
        await fs_1.fs.writeFile(plist, newPlist);
    }
    catch (err) {
        throw new Error(`Could not save plist: ${err.message}`, { cause: err });
    }
    if (!quiet) {
        logger_1.default.debug(`Wrote plist file '${plist}'`);
    }
}
/**
 * Creates a binary plist Buffer from an object
 *
 * @param data - The object to be turned into a binary plist
 * @returns Plist in the form of a binary buffer
 */
function createBinaryPlist(data) {
    return (0, bplist_creator_1.default)(data);
}
/**
 * Parses a Buffer into an Object
 *
 * @param data - The buffer of a binary plist
 * @returns Array of parsed root objects (typically one element)
 */
function parseBinaryPlist(data) {
    return (0, bplist_parser_1.parseBuffer)(data);
}
function createPlist(object, binary = false) {
    return binary ? createBinaryPlist(object) : (0, plist_1.build)(object);
}
/**
 * Parses a buffer or string into a JS object
 *
 * @param data - The plist as a string, Buffer, Uint8Array, or ArrayBuffer
 * @returns Parsed plist as a JS object
 * @throws Will throw an error if the plist type is unknown
 */
function parsePlist(data) {
    if (typeof data === 'string') {
        return data.startsWith(BPLIST_IDENTIFIER.TEXT)
            ? parseBinaryPlistRoot(Buffer.from(data))
            : (0, plist_1.parse)(data);
    }
    const binaryLikeData = toBufferIfBinaryLike(data);
    if (binaryLikeData) {
        return BPLIST_IDENTIFIER.BUFFER.compare(binaryLikeData, 0, BPLIST_IDENTIFIER.BUFFER.length) === 0
            ? parseBinaryPlistRoot(binaryLikeData)
            : (0, plist_1.parse)(binaryLikeData.toString());
    }
    throw new Error(`Unknown type of plist, data: ${(0, util_1.truncateString)(String(data), { length: 200 })}`);
}
function toBufferIfBinaryLike(data) {
    if (Buffer.isBuffer(data)) {
        return data;
    }
    if (data instanceof Uint8Array) {
        return Buffer.from(data.buffer, data.byteOffset, data.byteLength);
    }
    if (data instanceof ArrayBuffer) {
        return Buffer.from(data);
    }
    return null;
}
function parseBinaryPlistRoot(data) {
    const parsed = parseBinaryPlist(data);
    if (!parsed.length) {
        throw new Error(`Binary plist appears to be empty`);
    }
    return parsed[0];
}
